espressostreamer/espresso_streamer.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/solgen/go/bridgegen"
+	"github.com/offchainlabs/nitro/solgen/go/precompilesgen"
+	"github.com/offchainlabs/nitro/util/headerreader"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+type MessageWithMetadataAndPos struct {
+	MessageWithMeta arbostypes.MessageWithMetadata
+	Pos             uint64
+	HotshotHeight   uint64
+}
+
+type EspressoStreamer struct {
+	stopwaiter.StopWaiter
+	l1Reader                      *headerreader.HeaderReader
+	espressoClient                *espressoClient.MultipleNodesClient
+	nextHotshotBlockNum           uint64
+	currentMessagePos             uint64
+	namespace                     uint64
+	retryTime                     time.Duration
+	pollingHotshotPollingInterval time.Duration
+	messageWithMetadataAndPos     []*MessageWithMetadataAndPos
+	espressoTEEVerifierABI        *abi.ABI
+	espressoTEEVerifierAddr       common.Address
+}
+
+func NewEspressoStreamer(namespace uint64, hotshotUrls []string,
+	nextHotshotBlockNum uint64,
+	retryTime time.Duration,
+	pollingHotshotPollingInterval time.Duration,
+	parentChainNodeUrl string,
+	headerReaderConfig headerreader.Config,
+	espressoTEEVerifierAddress common.Address,
+) *EspressoStreamer {
+
+	espressoTEEVerifierAbi, err := bridgegen.IEspressoTEEVerifierMetaData.GetAbi()
+	if err != nil {
+		log.Crit("Unable to find EspressoTEEVerifier ABI")
+	}
+	l1Client, err := ethclient.Dial(parentChainNodeUrl)
+	if err != nil {
+		log.Crit("Failed to create l1 client", "url", parentChainNodeUrl)
+		return nil
+	}
+
+	arbSys, err := precompilesgen.NewArbSys(types.ArbSysAddress, l1Client)
+	if err != nil {
+		log.Crit("Failed to create arbsys", "err", err)
+		return nil
+	}
+
+	// we initialze a l1 reader that will poll for header every 60 seconds
+	l1Reader, err := headerreader.New(context.Background(), l1Client, func() *headerreader.Config {
+		return &headerReaderConfig
+	}, arbSys)
+
+	if err != nil {
+		log.Crit("Failed to create l1 reader", "err", err)
+		return nil
+	}
+	return &EspressoStreamer{
+		espressoClient:                espressoClient.NewMultipleNodesClient(hotshotUrls),
+		nextHotshotBlockNum:           nextHotshotBlockNum,
+		retryTime:                     retryTime,
+		pollingHotshotPollingInterval: pollingHotshotPollingInterval,
+		namespace:                     namespace,
+		l1Reader:                      l1Reader,
+		espressoTEEVerifierABI:        espressoTEEVerifierAbi,
+		espressoTEEVerifierAddr:       espressoTEEVerifierAddress,
+	}
+}
+
+func (s *EspressoStreamer) Reset(currentMessagePos uint64, currentHostshotBlock uint64) {
+	s.currentMessagePos = currentMessagePos
+	s.nextHotshotBlockNum = currentHostshotBlock
```

Comment:
I think we also need to clear the queue here. Otherwise the subsequent hotshot transactions will be regarded as duplicates.